GMnetENGINE.gmx/scripts/udphp_clientPunch.gml

```diff
@@ -63,9 +63,59 @@ if (!client_connected) {
         //Send a packet to the server to punch the hole. If this reaches the server, he will
         //add us to the list of players and send an answer.
         buffer_seek(client_buffer, buffer_seek_start, 0);
-        buffer_write(client_buffer, buffer_s8, udphp_packet.KNOCKKNOCK );
-        network_send_udp( client_udp, client_serverip, client_serverport, client_buffer, buffer_tell(client_buffer) );
-        if (client_timeout > global_timeout) {
+        buffer_write(client_buffer, buffer_s8, udphp_packet.KNOCKKNOCK );        
+        // Handle punch states
+        udphp_punchstate(client_id,client_udp,client_serverip,client_serverport,client_buffer);
+        // Change punch stages depending on the total timeout
+        if (client_timeout = floor(global_timeout*0.25) and global.udphp_punch_stage!="Try sequence external server port")
+        {
+            // Try change the external server port
+            // If the server NAT changed the port to a sequencent port nearby the received port from master server
+            // Some NAT change the external port when external ip change in the send network message.
+            // Reset some variables we use
+            global.udphp_punch_stage_sub1="";
+            global.udphp_punch_stage_counter=0;            
+            // Only config this once
+            global.udphp_punch_stage="Try sequence external server port";            
+        }   
+        else if (client_timeout = floor(global_timeout*0.40) and global.udphp_punch_stage!="Try server port or provided server port")
+        {
+            // Try connect with provided port and hope the server got portforward
+            // The port may be provided by the server to the master and to us if
+            // global.htme_object.udphp_provide_server_port=true in config
+            // Check if port is provided
+            if global.htme_object.server_port>0

```

Comment:
I know there was something fishy about this. I think that when I tested direct connect I run into problems with the port. But I don't remember what it was. We will see that later on when testing it more.